File: src/game/traffic.js
import { bridgePoint, bridgeGeometry, clearPosition } from "./geometry.js";
const worlds = new WeakMap();
// Narrow decks admit one direction at a time. Wide decks use two spatial lanes.
// This is derived traffic state; it is rebuilt from bodies after a rewind.
export function canEnterBridge(w, c, destination) {
  let traffic = worlds.get(w);
  if (!traffic) {
    traffic = new Map();
    worlds.set(w, traffic);
  }
  for (const bridge of w.objects) {
    if (bridge.type !== "bridge") continue;
    const g = bridgeGeometry(bridge);
    if (!g.complete || g.width >= 2) continue;
    const at = bridgePoint(bridge, c),
      to = bridgePoint(bridge, destination);
    if (at.t > 0.02 && at.t < 0.98 && at.distance < g.width / 2) return true;
    if (at.t < 0.5 === to.t < 0.5) continue;
    let record = traffic.get(bridge.id);
    const registered = record?.waiting.has(c.id) || false;
    const entry = at.t < 0.5 ? g.a : g.b;
    if (!registered && Math.hypot(c.x - entry.x, c.y - entry.y) > 6) continue;
    if (!record) {
      record = { waiting: new Map(), direction: null, until: 0 };
      traffic.set(bridge.id, record);
    }
    const direction = at.t < 0.5 ? 1 : -1;
    const waiting = record.waiting.get(c.id);
    if (!waiting) {
      record.waiting.set(c.id, { direction, since: w.time, lastSeen: w.time });
    } else if (waiting.direction !== direction) {
      record.waiting.set(c.id, { direction, since: w.time, lastSeen: w.time });
    } else {
      waiting.lastSeen = w.time;
    }
    for (const [id, value] of record.waiting)
      if (w.time - value.lastSeen >= 1) record.waiting.delete(id);
    const onDeck = w.creatures.some((other) => {
      const p = bridgePoint(bridge, other);
      return p.t > 0.02 && p.t < 0.98 && p.distance < g.width / 2;
    });
    if (onDeck) {
      const allowed = record.direction === direction && w.time < record.until;
      if (allowed) record.waiting.delete(c.id);
      return allowed;
    }
    if (w.time >= record.until || record.direction == null) {
      const first = [...record.waiting.entries()].sort(
        (a, b) => a[1].since - b[1].since || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0),
      )[0];
      record.direction = first?.[1].direction || direction;
      record.until = w.time + 4;
    }
    if (record.direction === direction) {
      record.waiting.delete(c.id);
      return true;
    }
    return false;
  }
  return true;
}

// Kept as an internal export so simulation travel can move queued residents
// clear of a narrow bridge entrance without exposing a new engine operation.
export function bridgeWaitingPoint(w, c) {
  for (const bridge of w.objects) {
    const waiting = worlds.get(w)?.get(bridge.id)?.waiting.get(c.id);
    if (!waiting) continue;
    const g = bridgeGeometry(bridge), { dx, dy } = bridgePoint(bridge, c);
    const sign = waiting.direction;
    const entry = sign === 1 ? g.a : g.b;
    const delta = { x: c.x - entry.x, y: c.y - entry.y };
    const back = Math.max(-sign * (delta.x * dx + delta.y * dy), 2.4);
    const side = Math.max(sign * (-delta.x * dy + delta.y * dx), 1.2);
    const point = {
      x: entry.x - dx * sign * back - dy * sign * side,
      y: entry.y - dy * sign * back + dx * sign * side,
    };
    return clearPosition(w, point) ? point : null;
  }
  return null;
}

File: engine/src/traffic.rs
use crate::{
    Engine,
    geometry::{Object, bridge, bridge_point},
    value::*,
};
use serde_json::Value;
use std::collections::HashMap;
#[derive(Default)]
pub(crate) struct BridgeTraffic {
    // Direction, first request, most recent request. Cancelled jobs must not
    // retain a turn indefinitely just because their resident is still alive.
    pub waiting: HashMap<String, (i8, f64, f64)>,
    pub direction: Option<i8>,
    pub until: f64,
}
impl Engine {
    pub(crate) fn can_enter_bridge(&mut self, c: &Value, destination: Point) -> bool {
        let time = num(&self.world, "time");
        for o in list(&self.world, "objects") {
            if text(o, "type") != "bridge" {
                continue;
            }
            let g = bridge(&Object::read(o));
            // Passing lanes are only created for decks at least two units wide.
            if !g.complete || g.width >= 2. {
                continue;
            }
            let at = bridge_point(g, Point::read(c));
            let to = bridge_point(g, destination);
            if at.0 > 0.02 && at.0 < 0.98 && at.1 < g.width / 2. {
                return true;
            }
            if (at.0 < 0.5) == (to.0 < 0.5) {
                continue;
            }
            let registered = self
                .sim
                .traffic
                .get(text(o, "id"))
                .is_some_and(|r| r.waiting.contains_key(text(c, "id")));
            if !registered && Point::read(c).distance(if at.0 < 0.5 { g.a } else { g.b }) > 6. {
                continue;
            }
            let record = self
                .sim
                .traffic
                .entry(text(o, "id").to_string())
                .or_default();
            let direction = if at.0 < 0.5 { 1 } else { -1 };
            record
                .waiting
                .entry(text(c, "id").to_string())
                .and_modify(|v| {
                    if v.0 != direction {
                        *v = (direction, time, time);
                    } else {
                        v.2 = time;
                    }
                })
                .or_insert((direction, time, time));
            record.waiting.retain(|_, v| time - v.2 < 1.);
            let on_deck = list(&self.world, "creatures").iter().any(|c| {
                let p = bridge_point(g, Point::read(c));
                p.0 > 0.02 && p.0 < 0.98 && p.1 < g.width / 2.
            });
            if on_deck {
                let allowed = record.direction == Some(direction) && time < record.until;
                if allowed {
                    record.waiting.remove(text(c, "id"));
                }
                return allowed;
            }
            if time >= record.until || record.direction.is_none() {
                record.direction = record
                    .waiting
                    .iter()
                    .min_by(|(a, (_, at, _)), (b, (_, bt, _))| at.total_cmp(bt).then(a.cmp(b)))
                    .map(|(_, v)| v.0)
                    .or(Some(direction));
                record.until = time + 4.;
            }
            if record.direction == Some(direction) {
                record.waiting.remove(text(c, "id"));
                return true;
            }
            return false;
        }
        true
    }

    /// A waiting queue stands beside the approach, leaving the exit usable.
    /// This is a walking destination, never a relocation or a saved reservation.
    pub(crate) fn bridge_waiting_point(&mut self, c: &Value) -> Option<Point> {
        let id = text(c, "id");
        let mut holding = None;
        for o in list(&self.world, "objects") {
            let Some(record) = self.sim.traffic.get(text(o, "id")) else {
                continue;
            };
            let Some(&(direction, _, _)) = record.waiting.get(id) else {
                continue;
            };
            let g = bridge(&Object::read(o));
            let (_, _, _, dx, dy) = bridge_point(g, Point::read(c));
            let sign = f64::from(direction);
            let entry = if direction == 1 { g.a } else { g.b };
            let delta = Point::read(c).offset(-entry.x, -entry.y);
            // Clear the approach without sending the tail of a large queue
            // arbitrarily far backwards into occupied workplaces.
            let back = (-sign * (delta.x * dx + delta.y * dy)).max(2.4);
            let side = (sign * (-delta.x * dy + delta.y * dx)).max(1.2);
            holding = Some(entry.offset(
                -dx * sign * back - dy * sign * side,
                -dy * sign * back + dx * sign * side,
            ));
            break;
        }
        holding.filter(|p| self.clear_position(*p))
    }
}
